python/ray/data/datasource/partitioning.py

```diff
@@ -0,0 +1,378 @@
+        >>> PartitionStyle(json.loads('"hive"'))  # PartitionStyle.HIVE
+    """
+
+    HIVE = "hive"
+    DIRECTORY = "dir"
+
+
+@DeveloperAPI
+class PathPartitionBase:
+    """Path-based partition formats embed all partition keys and values directly in
+    their dataset file paths.
+
+    Base class for path-based partition formats.
+    """
+
+    def __init__(
+        self,
+        style: PartitionStyle,
+        base_dir: Optional[str],
+        field_names: Optional[List[str]],
+    ):
+        """Creates a new path-based dataset partition scheme.
+
+        Args:
+            style: The partition style - may be either HIVE or DIRECTORY.
+            base_dir: "/"-delimited base directory that all partitioned paths should
+                exist under (exclusive).
+            field_names: The partition key field names (i.e. column names for tabular
+                datasets). Required when parsing DIRECTORY partitioned paths or
+                generating HIVE partitioned paths.
+        """
+        self._style = style
+        self._base_dir = base_dir or ""
+        self._normalized_base_dir = None
+        self._field_names = field_names
+
+    @property
+    def style(self) -> PartitionStyle:
+        """Gets the path partitioning style."""
+        return self._style
+
+    @property
+    def base_dir(self) -> str:
+        """Gets the base directory."""
+        return self._base_dir
+
+    @property
+    def normalized_base_dir(self) -> Optional[str]:
+        """Returns the base directory normalized for compatibility with a filesystem.
+
+        The normalized base directory is undefined by default, then lazily resolved
+        or updated by any operation that takes a filesystem as input.
+        """
+        return self._normalized_base_dir
+
+    @property
+    def field_names(self) -> Optional[List[str]]:
+        """Gets the partition key field names."""
+        return self._field_names
+
+    def _normalize_base_dir(self, filesystem: "pyarrow.fs.FileSystem"):
+        """Normalizes the partition base directory for compatibility with the
+        given filesystem.
+
+        This should be called once a filesystem has been resolved to ensure that this
+        base directory is correctly discovered at the root of all partitioned file
+        paths.
+        """
+        from ray.data.datasource.file_based_datasource import (
+            _resolve_paths_and_filesystem,
+        )
+
+        paths, _ = _resolve_paths_and_filesystem(
+            self._base_dir,
+            filesystem,
+        )
+        assert (
+            len(paths) == 1
+        ), f"Expected 1 normalized base directory, but found {len(paths)}"
+        normalized_base_dir = paths[0]
+        if len(normalized_base_dir) and not normalized_base_dir.endswith("/"):
+            normalized_base_dir += "/"
+        self._normalized_base_dir = normalized_base_dir
+
+
+@DeveloperAPI
+class PathPartitionGenerator(PathPartitionBase):
```

Comment:
My understanding of roles for these 3 classes is that
- `PathPartitionBase` (or its subclass - do we need that?) describes a path-based partition
- `PathPartitionGenerator` encodes the path-based partition (into string): so it seems not a IS-A relationship, hence not a good fit to model as a subclass
- `PathPartitionParser` decodes a path-based partition (out of string), which can then be used for use cases like filtering: so similar it's not a IS-A relationship
So it looks we should model them with composition not inheritance. WDYT?

Reply:
We can discuss this, but developing on top of the above, I think we can create abstractions and their roles:
- PathPartition: describes a path-based partition
- PathPartitionEncoder: encodes the path-based partition into string, so we can pass it around and use it in writing out Dataset to file systems
- PathPartitionParser (or PathPartitionDecoder): decodes a path-based partition out of string, so we can easily access each fields/values, and use them for e.g. filtering
- PathPartitionSelector (or PathPartitionFilter): given the PathPartition, the selector/filter function, and the field/values, it produces a subset of partitions (to read into Dataset)

Btw, we can have an offline meeting if this is a bit complex or inefficient to discuss over GitHub :)

Reply:
I definitely like the cleaner separation of responsibilities that composition provides vs. inheritance. Let me know what you think of the latest refactor here. One naming difference is that I explicitly renamed `PathPartitionBase` to `PathPartitionScheme` instead of `PathPartition`, since I felt like the latter gave the impression that it was only referring to one partition, when in fact it holds the spec pertaining an arbitrarily large number of partitions.

One thing I didn't like initially was that constructing something like a `PathPartitionFilter` required first constructing a `PathPartitionParser` and a `PathPartitionScheme`, but decided to strike a balance here with static factories that provide alternate constructors with flattened arguments.

Reply:
Thanks! For the naming, the PathPartitionScheme looks great!

What do you think making PathPartitionFilter as a function v.s. as a class?